### DISCUSSION TITLE
@babel/traverse "Couldn't find a Program"

### DISCUSSION THREAD

**Discussion by npavlovichero** [babel/babel#13742] [Q&A] [answered]
Hello. Could someone help me shed some light as to what this error is exactly?

**Comment by nicolo-ribaudo** [comment, marked as answer]
Where are you getting this error from? Are you using `@babel/traverse` directly? It usually means that you are trying to traverse a subset of the AST, but the main traverse needs a full AST (starting from a node with `type: "Program"`) to perform scope analysis.

**Reply by npavlovichero** [reply]
I'm not using anything directly. ESLint breaks.

**Reply by nicolo-ribaudo** [reply]
ESLint doesn't rely on `@babel/traverse`. Can you share a repository which shows the error?

**Reply by npavlovichero** [reply]
![image](https://user-images.githubusercontent.com/89974341/140517238-b2a71b04-53e5-41c4-9b91-8bdc2a9720e9.png)
```
Error: Couldn't find a Program
Occurred while linting /*hidden*/*/index.tsx:661

Error: Couldn't find a Program
Occurred while linting /*hidden*/*/index.tsx:661
    at Scope.getProgramParent (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/scope/index.js:817:11)
    at Scope.crawl (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/scope/index.js:722:32)
    at Scope.init (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/scope/index.js:711:12)
    at NodePath.setScope (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/path/context.js:132:30)
    at NodePath.setContext (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/path/context.js:148:8)
    at NodePath.pushContext (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/path/context.js:214:8)
    at TraversalContext.visitQueue (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/context.js:89:14)
    at TraversalContext.visitSingle (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/context.js:73:19)
    at TraversalContext.visit (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/context.js:127:19)
    at Function.traverse.node (#REPOSITORY_ROOT#/node_modules/@babel/traverse/lib/index.js:76:17)
Process finished with exit code -1
```

**Reply by npavlovichero** [reply]
My guess is, judging by your first reply and this, that it's probable that whatever is using it silently fails at building AST then tries traversing it without any prior validation.

**Reply by npavlovichero** [reply]
Yup, hello eslint-plugin-react-native. Okay @nicolo-ribaudo thanks for clearing that up, made pinpointing this much easier.

**Reply by billnbell** [reply]
what was it?

**Reply by ithustle** [reply]
How to fix that?

**Reply by JLHwung** [reply]
This is a downstream bug: https://github.com/Intellicode/eslint-plugin-react-native/pull/314. Before fix is landed, you can disable the `react-native/no-raw-text` rule.

**Reply by ithustle** [reply]
Hi @JLHwung . Please, find bellow my `.eslintrc`:
```
module.exports = {
  root: true,
  extends: ['@react-native-community', 'plugin:react-native/all'],
  parser: '@typescript-eslint/parser',
  plugins: ['@typescript-eslint', 'react', 'react-native'],
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    }
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx'],
      rules: {
        '@typescript-eslint/no-shadow': ['error'],
        'no-shadow': 'off',
        'no-undef': 'off',
      },
    },
  ],
};
```
I have no `react-native/no-raw-text` rule

**Reply by JLHwung** [reply]
> extends: ['plugin:react-native/all'],

I think that preset includes all `react-native/*` rules.

**Comment by tichopad** [comment]
Hijacking this thread, since it's the first to come up in search results.

I have a monorepo setup with a React Native (Expo) app and I got the error while trying to run the app locally after upgrading Expo SDK.
For me, it's been caused by multiple equivalent version ranges in a lockfile pointing to two different versions of @babel/* packages (7.18.x and 7.19.x).
Deduplication fixed it. In my case, with Yarn, that's done by running `yarn dedupe @babel/*`.

**Comment by marcus-pousette** [comment]
I am not sure whether this is related or not but I got this in a CRA5 app and after some long time digging... I solved this with

Changing package.json 
```
 "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
  ```
  to 
  ```
   "browserslist": {
    "production": [
      "chrome >= 67",
      "edge >= 79",
      "firefox >= 68",
      "opera >= 54",
      "safari >= 14"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
  ```

**Comment by ospfranco** [comment]
Hey I started facing this issue when updating to the latest version of RN (0.71), it doesn't seem to be related to eslint in my case.

![dCTuV](https://user-images.githubusercontent.com/1634213/213387714-9cea0b8c-49d2-45d3-804e-4246a534c9d8.jpg)

You can find the complete info on my [StackOverflow question](https://stackoverflow.com/questions/75169132/react-native-with-typeorm-babel-transform-error-couldnt-find-program)

**Reply by JLHwung** [reply]
Did you try removing the npm/yarn lock files and install again?

**Reply by ospfranco** [reply]
yep, I nuked everything. Also tried deduping babel. No success.

**Reply by XantreDev** [reply]
fixed this issue for me
```json
  "resolutions": {
    "@babel/traverse": "7.21.3",
    "@babel/core": "^7.21.3"
  }
```

**Reply by Jacksonishere** [reply]
Hi, couple months late here but did you end up figuring out a solution?

**Reply by ospfranco** [reply]
No, I tried the resolutions but it didn't work. I also left this months ago, maybe it is worth trying again with the latest versions.

**Comment by DanyPell** [comment]
`Error: /var/www/myProject/pages/api/users/deleteUser.ts: Couldn't find a Program`

Seemingly out of nowhere.

**Comment by cth166** [comment]
The error happens when parentPath fails to locate a Program-type node.

- If you're just trying to modify node properties and used a manually created dummyNode, make sure to include noScope: true in your visitors.

- If you need to properly emulate scope chain behavior, you should construct the node with t.file instead.

```
import type { Declaration, Identifier } from '@babel/types';
import * as t from '@babel/types';
import traverse from "@babel/traverse";

function removeThis(node: Declaration) {
    traverse(getDummyAst(node), {
        MemberExpression(path) {
            const { node } = path;
            if (t.isThisExpression(node.object)) {
                const identifier = t.identifier((node.property as Identifier).name);
                path.replaceWith(identifier);
            }
        },
        noScope: true,
    });
}
function getDummyAst(node: Declaration) {
    return t.file(t.program([node]));
}
```

**Comment by LUCIA123-BEEP** [comment]
Yo solo pase por aqui alguin me pude enseñar codigo morse?
